### PR TITLE
New function datasources no longer cause datasource change internally

### PR DESCRIPTION
**Pull request by tomivirkki**
[Review on Reviewable](https://reviewable.io/reviews/vaadin/components/84)

### CODE REVIEW

**Comment by tepi** [comment]
Reviewed 2 of 2 files at r1.
Review status: :shipit: all files reviewed at latest revision, all discussions resolved, all commit checks successful.

---



Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/vaadin/components/84)

File: vaadin-components-gwt/src/main/java/com/vaadin/components/grid/GridComponent.java [hunk 1]
```diff
@@ -268,7 +268,13 @@ public void setHeight(String height) {
 
     public void setDataSource(JavaScriptObject data) {
         if (JsUtils.isFunction(data)) {
-            grid.setDataSource(new GridJsFuncDataSource(data, this));
+            if (grid.getDataSource() instanceof GridJsFuncDataSource) {
+                ((GridJsFuncDataSource) grid.getDataSource())
+                        .setJSFunction(data);
+                getSelectionModel().reset();
```
Comment (manolo) [inline vaadin-components-gwt/src/main/java/com/vaadin/components/grid/GridComponent.java]: Shouldn't this be done in the new setJsFunction method?

**Comment by manolo** [comment]
LGTM, but I don't see a big performance improvement, but anyway, but any minor improvement counts.

**Comment by tomivirkki** [comment]
The performance improvement especially stands out when the function datasources have the same count of items (or lots of items); like in the example app of the guy who reported the Github issue. The grid in his app used to flicker every time the data changed (he always set a new ds with grid.data.source = [...]) but with the new version there's no flickering at all.

---







Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/vaadin/components/84)

**Comment by tomivirkki** [comment]
Review status: 0 of 2 files reviewed at latest revision, 1 unresolved discussion, some commit checks pending.

---

<sup>**[vaadin-components-gwt/src/main/java/com/vaadin/components/grid/GridComponent.java, line 274 \[r1\]](https://reviewable.io:443/reviews/vaadin/components/84#-JwuhqxUWmRe6WxwPOkP-r1-274)** ([raw file](https://github.com/vaadin/components/blob/fdb4d5008b95f47df615b52059bb53c2c468bf17/vaadin-components-gwt/src/main/java/com/vaadin/components/grid/GridComponent.java#L274)):</sup>
Done.

---


Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/vaadin/components/84)

**Pull request merged**